File: app/src/main/java/com/example/hydrogelsimulator/FirebaseHelper.java
package com.example.hydrogelsimulator;

import com.google.firebase.firestore.FirebaseFirestore;

import java.util.HashMap;
import java.util.Map;

/** Persists every simulation record to Firestore (collection: "simulations"). */
public class FirebaseHelper {

    private final FirebaseFirestore db = FirebaseFirestore.getInstance();

    public void saveSimulation(HydrogelPredictor p, String aiNotes) {
        Map<String, Object> data = new HashMap<>();
        data.put("gelatin",        p.gelatin);
        data.put("genipin",        p.genipin);
        data.put("pH",             p.pH);
        data.put("temperature",    p.temp);
        data.put("tensileStrength",p.tensileStrength());
        data.put("elasticity",     p.elasticity());
        data.put("degradationDays",p.degradationDays());
        data.put("swellingRatio",  p.swellingRatio());
        data.put("stability",      p.stabilityScore());
        data.put("recommendation", p.recommendation());
        data.put("aiNotes",        aiNotes == null ? "" : aiNotes);
        data.put("timestamp",      System.currentTimeMillis());

        db.collection("simulations").add(data);
    }
}

File: flutter_app/lib/services/firebase_helper.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';

import '../models/hydrogel_predictor.dart';

class FirebaseHelper {
  Future<void> saveSimulation(HydrogelPredictor p, String aiNotes) async {
    try {
      await FirebaseFirestore.instance.collection('simulations').add({
        'gelatin': p.gelatin,
        'genipin': p.genipin,
        'pH': p.pH,
        'temperature': p.temp,
        'tensileStrength': p.tensileStrength(),
        'elasticity': p.elasticity(),
        'degradationDays': p.degradationDays(),
        'swellingRatio': p.swellingRatio(),
        'stability': p.stabilityScore(),
        'recommendation': p.recommendation(),
        'aiNotes': aiNotes,
        'timestamp': FieldValue.serverTimestamp(),
      });
    } catch (e) {
      debugPrint('Firestore save failed: $e');
    }
  }
}
